### ISSUE TITLE
Alternative syntax for `map(func, x)`

### ISSUE THREAD

**Issue by kmsquire** [JuliaLang/julia#8450]
This was discussed in some detail [here](https://github.com/JuliaLang/julia/issues/8389#issuecomment-55934489).  I was having trouble finding it, and thought it deserved its own issue.

**Comment by JeffBezanson** [comment]
I like the `f(x) over x` idea. We could even use `f(x) for x` to avoid a new keyword. In fact `[f(x) for x=x]` already works. We would then need to make comprehensions equivalent to `map`, so they can work for non-Arrays. `Array` would just be the default.

**Comment by johansigfrids** [comment]
I must admit I've also come to like the `over` idea. One difference between `over` as map and `for` in list comprehension is what happens in case of multiple iterators: `[f(x, y) for x=x, y=y]` results in a matrix. For the map case you generally still want a vector, is `[f(x, y) over x, y]` would be equivalent to `[f(x, y) for (x,y) = zip(x, y)]]`. Because of this I still think introducing an additional keyword `over` is worth it, because, as this issuer has raised, `map`ing over multiple vectors is very common and needs to be terse.

**Comment by nalimilan** [comment]
Hey, I convinced Jeff about the syntax! ;-)

**Comment by JeffBezanson** [comment]
This belongs beside #4470, so adding to 0.4-projects for now.

**Comment by krcz** [comment]
If I understand the gist of discussion the main problem is that we want to get mapping-like syntax that:
* works with various data types, like DataArrays, not only native arrays;
* is as fast as manually written loop.

It might be possible to do it using inlining, but being really careful to make sure inlining works.

What about different approach: using macro depending on inferred data type. If we can infer that data structure is DataArray, we use map-macro provided by DataArrays library. If it is SomeKindOfStream, we use stream library provided one. If we cannot infer type, we just use dynamically dispatched general implementation.

This might force data structures creators to write such macros, but it would be needed only if its author wanted it to have really efficient execution.

If what I'm writing is unclear, I mean something like `[EXPR for i in collection if COND]` could be translated to `eval(collection_mapfilter_macro(:(i), :(EXPR), :(COND)))`, where `collection_mapfilter_macro` is chosen based on inferred collection type.

**Comment by JeffBezanson** [comment]
No, we don't want to do things like that. If DataArray defines `map` (or the equivalent), its definition should always be called for DataArrays regardless of what can be inferred.

This issue actually isn't about implementation, but syntax. Right now many people are used to `sin(x)` implicitly mapping if `x` is an array, but there are many problems with that approach. The question is what alternate syntax would be acceptable.

**Comment by Mike43110** [comment]
1) Support for allocation-free in-place assignment using  `y[:] = sqrt.(x .+ 2) `
2) Support for allocation-free reductions like  `sum((x .- y).^2, 1) `

y =  √π exp(-x^2) * sin(k*x) + im * log(x-1)

Looking at these three examples from others I think with the `for` syntax this would end up something like this:
1) `y[:] = [ sqrt(x + 2) for x ])`
2) `sum([ (x-y)^2 for x,y ], 1)`
and
`y = [ √π exp(-x^2) * sin(k*x) + im * log(x-1) for x,k ]`

I like this quite a lot! The fact it creates a temporary array is quite explicit and it is still readable and terse.

Minor question though, could `x[:] = ... for x` have some magic to mutate the array?
I am not sure if this would provide much benefit but I can imagine it would help for large arrays.
I can believe though that it may be a completely different kettle of fish which should be discussed someplace else.

**Comment by nalimilan** [comment]
@Mike43110 Your `x[:] = [ ... for x ]` could be written `x[:] = (... for x)`, the RHS creating a generator, which would be collected element by element to fill `x`, without allocating a copy. That was the idea behind my `LazyArray` experiment above.

**Comment by mschauer** [comment]
The ``[f <- y]`` syntax would be nice if combined with a ``Int[f <- y]`` syntax for a map which knows its output type and does not need to interpolate from ``f(y[1])`` what the other elements will be.

**Comment by mschauer** [comment]
Especially, as this gives an intuitive interface for ``mapslices`` as well,   ``[f <- rows(A)]`` where ``rows(A)`` (or ``columns(A)`` or ``slices(A, dims)``) returns a ``Slice`` object so dispatch can be used:
```
map(f, slice::Slice) = mapslices(f, slice.A, slice.dims)
```

**Comment by JeffBezanson** [comment]
When you add indexing, this gets a bit harder. For example

```
f(x[:,j]) .* g(x[i,:])
```

It's hard to match the conciseness of that. The blowup from comprehension style is pretty bad:

```
[f(x[m,j])*g(x[i,n]) for m=1:size(x,1), n=1:size(x,2)]
```

where, to make matters worse, cleverness was required to know that this is a case of nested iteration, and can't be done with a single `over`.

**Comment by nalimilan** [comment]
@JeffBezanson I'm not sure what's the intention of your comment. Has anybody suggested to get rid of the `.*` syntax?

**Comment by JeffBezanson** [comment]
No; I'm focusing on `f` and `g` here. This is an example where you can't just add `over x` at the end of the line.

**Comment by nalimilan** [comment]
OK, I see, I had missed the end of the comment. Indeed the dots version is nicer in that case.

Though with array views, there will be a reasonably efficient (AFAICT) and not so ugly alternative:
``[ f(y) * g(z) for y in x[:,j], z in x[i,:] ]``

**Comment by jdbates** [comment]
Could the above example be solved by nesting over keywords?

`f(x)*g(y) over x,y`

is interpreted as

`[f(x)*g(y) for (x,y) = zip(x,y)]`

whereas

`f(x)*g(y) over x over y`

becomes

`[f(x)*g(y) for x=x, y=y]`

**Comment by davidagold** [comment]
@JeffBezanson How about

`f(x[:i,n]) * g(x[m,:i]) over i`

gives the equivalent of `f.(x[:,n] .* g.(x[m,:])`. The new syntax `x[:i,n]` means that `i` is being introduced locally as an iterator over the indices of the container `x[:,n]`. I don't know if that's possible to implement. But it seems (prima facie) neither ugly nor cumbersome, and the syntax itself gives bounds for the iterator, namely 1:length(x[:,n]). As far as keywords are concerned, "over" can signal that the entire range is to be used, whereas "for" can be used if the user wishes to specify a subrange of 1:length(x[:,n]):

`f(x[:i,n]) * g(x[m,:i]) for i in 1:length(x[:,n])-1`.

**Comment by StefanKarpinski** [comment]
@davidagold, `:i` already means the symbol `i`.

**Comment by davidagold** [comment]
Ah yes, good point. Well, as long as dots are fair game what about

`f(x[.i,n]) * g(x[m,.i]) over i`

where the dot indicates that `i` is being introduced locally as an iterator over 1:length(x[:,n). I suppose in essence this switches the dot notation from modifying the functions to modifying the arrays, or rather their indices. This would save one from the "dot creep" Jeff noted:

`[ f(g(e^(x[m,.i]))) * p(e^(f(y[.i,n]))) over i ]`

as opposed to

`f.(g.(e.^(x[m,:]))) .* p.(e.^(f.(y[:,n])))`

although I suppose the latter is slightly shorter. One potential advantage of the comprehension syntax is that it could allow for a wider range of patterns of element-wise operation. For instance, if the parser understood that indexing with `i` in `x[m, .i]` is implicitly modulo length(x[m,:]), then one could write

`[ f(x[.i]) * g(y[.j]) over i, j=-i ]`

To multiply the elements of `x` against the elements of `y` in the opposite order, i.e. the first element of `x` against the last element of `y`, etc. One could write

`[ f(x[.i]) * g(y[.j]) over i, j=i+1 ]`

to multiply the `i`th element of `x` by the `i+1`th element of `y` (where the last element of `x` would be multiplied by the first element of `y` due to indexing being understood in this context as modulo length(x)). And if `p::Permutation` permutes (1, ..., length(x)) one could write

`[ f(x[.i]) * g(y[.j]) over i, j=p(i) ]`

to multiply the `i`th element of `x` by the `p(i)`th element of `y`. 

Anyway, that's just a humble opinion from an outsider on an entirely speculative issue. =p I do appreciate the time anybody takes to consider it.

**Comment by bramtayl** [comment]
A souped up version of vectorize that will use r style recycling could be pretty useful. That is, arguments which do not match the size of the largest argument are extended via recycling. Then users can easily vectorize anything they want, regardless of the number of arguments etc.

    unvectorized_sum(a, b, c, d) = a + b + c + d
    vectorized_sum = @super_vectorize(unvectorized_sum)

    a = [1, 2, 3, 4]
    b = [1, 2, 3]
    c = [1, 2]
    d = 1

    A = [1, 2, 3, 4]
    B = [1, 2, 3, 1]
    C = [1, 2, 1, 2]
    D = [1, 1, 1, 1]

    vectorized_sum(a, b, c, d) = vectorized_sum(A, B, C, D) = [4, 7, 8, 7]

**Comment by johnmyleswhite** [comment]
I tend to think that recycling trades off too much safety for convenience. With recycling, it's very easy to write buggy code that executes without raising any errors.

**Comment by tkelman** [comment]
The first time I read about that behavior of R, I immediately wondered why someone would ever think that was a good idea. That's a really odd and surprising thing to do implicitly on mismatched-size arrays. There might be a handful of cases where it's how you would want to extend the smaller array, but just as well you might want zero-padding or repeating the end elements, or extrapolation, or any number of other application-dependent choices.

**Comment by bramtayl** [comment]
Whether or not to use @super_vectorize would be put in the hands of the user. It would also be possible to give warnings for various cases. For example, in R,

    c(1, 2, 3) + c(1, 2)
    [1] 2 4 4
    Warning message:
    In c(1, 2, 3) + c(1, 2) :
      longer object length is not a multiple of shorter object length

**Comment by tkelman** [comment]
I have no objections to making that an optional thing that users can choose whether or not to use, but it doesn't need to be implemented in the base language when it can just as well be done in a package.

**Comment by bramtayl** [comment]
`@vectorize_1arg` and `@vectorize_2arg` are both already included in Base, and the options they give a user seem somewhat limited.

**Comment by johnmyleswhite** [comment]
But this issue is focused on the design of a system for removing `@vectorize_1arg` and `@vectorize_2arg` from Base. Our goal is to remove vectorized functions from the language and to replace them with a better abstraction.

**Comment by JeffBezanson** [comment]
For example, recycling can be written as 

```
[ A[i] + B[mod1(i,length(B))] for i in eachindex(A) ]
```

which to me is pretty close to the ideal way to write it. Nobody needs to build it in for you. The main questions are (1) can this be made more concise, (2) how to extend it to other container types.

**Comment by Mike43110** [comment]
Looking at @davidagold's proposal I wondered if `var:` couldn't be used for that kind of thing where the variable would be the name before the colon. I saw this syntax used to mean `A[var:end]` so it does seem to be available.

`f(x[:,j]) .* g(x[i,:])` would then be `f(x[a:,j]) * g(x[i,b:]) for a, b` which isn't too much worse.

Multiple colons would be slightly weird though.

`f(x[:,:,j]) .* g(x[i,:,:])` -> `f(x[a:,a:,j]) * g(x[i,b:,b:]) for a, b` was my initial thought on this.

**Comment by bramtayl** [comment]
Ok, so here's a brief stab at a program for recycling. It would probably be possible to incorporate tuples analagously to vectors.

    using DataFrames

    a = [1, 2, 3]
    b = 1
    c = [1 2]
    d = @data [NA, 2, 3]

    argument_list = (a, b, c, d)

    recycle = function(argument_list...)

      # largest dimension in arguments
      max_dimension = maximum([ndims(i) for i in argument_list])
      # initialize dimension extents to 1
      dimension_extents = [1 for i in 1:max_dimension]

      # loop through argument and dimension
      for argument_index in 1:length(argument_list)
        for dimension in 1:ndims(argument_list[argument_index])
          # find the largest size for each dimension
          dimension_extents[dimension] = maximum([
            size(argument_list[argument_index], dimension),
            dimension_extents[dimension]
          ])
        end
      end

      # dataset which will specify necessary number of repmats for each argument
      dimension_ratios = [copy(dimension_extents) for i in 1:length(argument_list)]

      # loop through arguments and dimensions
      for argument_index in 1:length(argument_list)
        for dimension in 1:ndims(argument_list[argument_index])
          # find the minimum number of repmats until dataset is larger than the desired size
          dimension_ratios[argument_index][dimension] =
            ceil(dimension_extents[dimension] / size(argument_list[argument_index], dimension))
        end
      end

      # a dataset to hold the expanded arguments
      expand_arguments = Array(AbstractArray, length(argument_list))

      # build tuples of dimension ranges
      dimension_ranges = [1:i for i in dimension_extents]
      dimension_ranges = tuple(dimension_ranges...)

      # loop through arguments
      for argument_index in 1:length(argument_list)
        # repmat array arguments to desired size
        if typeof(argument_list[argument_index]) <: AbstractArray
          expand_arguments[argument_index] = 
            repmat(argument_list[argument_index], dimension_ratios[argument_index]...)
        # or, fill singletons to desired size         
        else
          expand_arguments[argument_index] = 
            fill(argument_list[argument_index], dimension_ratios[argument_index]...)
        end
        # now trim down data to the desired size
        expand_arguments[argument_index] = 
          getindex(expand_arguments[argument_index], dimension_ranges...)
      end
      expand_arguments
    end

    recycle(a, b, c, d)

    mapply = function(FUN, argument_list...)
      argument_list = recycle(argument_list...)
      FUN(argument_list...)
    end

    mapply(+, a, b, c, d)

Clearly, this isn't the most elegant or fast code (I'm a recent R immigrant). I'm not sure how to get from here to a @vectorize macro.

**Comment by StefanKarpinski** [comment]
You really don't want to define functions with the `foo = function(x,y,z) ... end` syntax in Julia, although it does work. That creates a non-constant binding of the name to an anonymous function. In Julia, the norm is to use generic functions and the bindings to functions are automatically constant. Otherwise you're going to get terrible performance.

**Comment by JeffBezanson** [comment]
I don't see why repmat is necessary here. Arrays filled with the index of each position are also a warning sign: it shouldn't be necessary to use a large chunk of memory to represent so little information. I believe such techniques are really only useful in languages where everything needs to be "vectorized". It seems to me the right approach is just to run a loop where some indexes are transformed, as in https://github.com/JuliaLang/julia/issues/8450#issuecomment-111898906.

**Comment by bramtayl** [comment]
Yes, that makes sense. Here's a start, but I'm having trouble figuring out how to do the looping at the end and then make a `@vectorize` macro.

```julia
function non_zero_mod(big::Number, little::Number)
  result = big % little
  result == 0 ? little : result
end

non_zero_mod(4, 3)

function mod_select(array, index...)
  # just return singletons
  if !(typeof(array) <: AbstractArray) return array end
  # find a new index with moded values
  transformed_index = 
      [non_zero_mod( index[i], size(array, i) )
       for i in 1:ndims(array)]
  # return value at moded index
  array[transformed_index...]
end

function mod_value_list(argument_list, index...)
  [mod_select(argument, index...) for argument in argument_list]
end

mapply = function(FUN, argument_list...)

  # largest dimension in arguments
  max_dimension = maximum([ndims(i) for i in argument_list])
  # initialize dimension extents to 1
  dimension_extents = [1 for i in 1:max_dimension]

  # loop through argument and dimension
  for argument_index in 1:length(argument_list)
    for dimension in 1:ndims(argument_list[argument_index])
      # find the largest size for each dimension
      dimension_extents[dimension] = maximum([
        size(argument_list[argument_index], dimension),
        dimension_extents[dimension]
      ])
    end
  end
  
  # more needed here
  # apply function over arguments using mod_value_list on arguments at each position
end
```

**Comment by ScottPJones** [comment]
In the talk, @JeffBezanson mentioned the syntax `sin(x) over x`, why not something more like:
`sin(over x)`? (or use some character instead of having `over` as a keyword)

**Comment by dphildebrandt** [comment]
Once this is solved we can also resolve #11872

**Comment by jrevels** [comment]
I hope I'm not late to the party, but I'd just like to offer a +1 to @davidagold's syntax proposal. It's conceptually clear, terse, and feels really natural to write. I'm not sure whether `.` would be the best identifying character, or how feasible an actual implementation would be, but one could make proof of concept using a macro to try it out (essentially like `@devec`, but might even be easier to implement).

**Comment by jrevels** [comment]
It also has the benefit of "fitting in" with existing array comprehension syntax:

```julia
result = [g(f(.i), h(.j)) over i, j]
```

vs.

```julia
result = [g(f(_i), h(_j)) for _i in eachindex(i), _j in eachindex(j)]
```

The key difference between the two being that the former would have more restrictions on shape-preservation, as it implies a map.

**Comment by mdcfrancis** [comment]
`over`, `range`, and `window` have some prior art in the OLAP space as modifiers to iteration, this seems consistent.

I'm not keen on the `.` syntax as that seems like a creep towards line noise. 

$ is perhaps consistent, intern the values of iterating i,j into the expression?
result = [g(f($i), h($j)) over i, j]

For automatic vectorization of an expression can we not `taint` one of the vectors in the expression and have the type system lift the expression into the vector space? 

I do similar to with time series operations where the expressiveness of julia already allows me to write 

ts_a = GetTS( ... )
ts_b = GetTS( ... ) 
factors = [ 1,  2, 3 ]

ts_x = ts_a * 2 + sin( ts_a * factors ) + ts_b 

which when observed outputs a time series of vectors. 

The main part missing is the ability to automatically lift existing functions into the space. This has to be done by hand

Essentially I'd like to be able to define something like the following ... 

abstract TS{K}
function {F}{K}( x::TS{K}, y::TS{K} ) = tsjoin( F, x, y ) 
# tsjoin is a time series iteration operator

and then be able to specialize for specific operations 

function mean{K}(x::TS{K}) = ... # my hand rolled form

**Comment by ma-laforge** [comment]
Hi @JeffBezanson,

If I understand correctly, I would like to propose a solution to your JuliaCon 2015 comment regarding a comment made above:
"[...] And telling library writers to put @vectorize on all appropriate functions is silly; you should be able to just write a function, and if somebody wants to compute it for every element they use map."
(But I will not address the other fundamental issue "[..] no really convincing reason why sin, exp etc. should be implicitly mapped over arrays.")

In Julia v0.40, I have been able to get a solution somewhat nicer (in my opinion) than @vectrorize:

    abstract Vectorizable{Fn}
    #Could easily have added extra argument to Vectorizable, but want to show inheritance case:
    abstract Vectorizable2Arg{Fn} <: Vectorizable{Fn}
    
    call{F}(::Type{Vectorizable2Arg{F}}, x1, x2) = eval(:($F($x1,$x2)))
    function call{F,T1,T2}(fn::Type{Vectorizable2Arg{F}}, v1::Vector{T1}, v2::Vector{T2})
        RT = promote_type(T1,T2) #For type stability!
        return RT[fn(v1[i],v2[i]) for i in 1:length(v1)]
    end
    
    #Function in need of vectorizing:
    function _myadd(x::Number, y::Number)
        return x+y+1
    end

    #"Register" the function as a Vectorizable 2-argument (alternative to @vectorize):
    typealias myadd Vectorizable2Arg{:_myadd}
    
    @show myadd(5,6)
    @show myadd(collect(1:10),collect(21:30.0)) #Type stable!

This is more-or-less reasonable, but is somewhat similar to the @vectorize solution.  In order for vectorization to be elegant, I suggest Julia support the following:

    abstract Vectorizable <: Function
    abstract Vectorizable2Arg <: Vectorizable
    
    function call{T1,T2}(fn::Vectorizable2Arg, v1::Vector{T1}, v2::Vector{T2})
        RT = promote_type(T1,T2) #For type stability!
        return RT[fn(v1[i],v2[i]) for i in 1:length(v1)]
    end

    #Note: by default, functions would normally be <: Function:
    function myadd(x::Number, y::Number) <: Vectorizable2Arg
        return x+y+1
    end

That's it! Having a function inherit from a Vectorizable function would make it vectorizable.

I hope this along the lines of what you were looking for.

Regards,

MA

**Comment by johnmyleswhite** [comment]
In the absence of multiple inheritance, how does a function inherit from `Vectorizable` and from something else? And how do you relate the inheritance information for specific methods to the inheritance information for a generic function?

**Comment by JeffBezanson** [comment]
@ma-laforge You can do that already --- define a type `myadd <: Vectorizable2Arg`, then implement `call` for `myadd` on `Number`.

**Comment by ma-laforge** [comment]
Thanks for that @JeffBezanson!

Indeed, I can almost my my solution look almost as good as what I want:

    abstract Vectorizable
    #Could easily have parameterized Vectorizable, but want to show inheritance case:
    abstract Vectorizable2Arg <: Vectorizable
    
    function call{T1,T2}(fn::Vectorizable2Arg, v1::Vector{T1}, v2::Vector{T2})
        RT = promote_type(T1,T2) #For type stability!
        return RT[fn(v1[i],v2[i]) for i in 1:length(v1)]
    end
    
    #SECTION F: Function in need of vectorizing:
    immutable MyAddType <: Vectorizable2Arg; end
    const myadd = MyAddType()
    function call(::MyAddType, x::Number, y::Number)
        return x+y+1
    end
    
    @show myadd(5,6)
    @show myadd(collect(1:10),collect(21:30.0)) #Type stable

Now, the only thing missing would be a way to "subtype" any function, so that that whole section F could be replaced with the more elegant syntax:

    function myadd(x::Number, y::Number) <: Vectorizable2Arg
        return x+y+1
    end

NOTE: I made the type "MyAddType", and the function name into a singleton object "myadd" because I find the resultant syntax nicer than if one were using `Type{Vectorizable2Arg}` in the call signature:

    function call{T1,T2}(fn::Type{Vectorizable2Arg}, v1::Vector{T1}, v2::Vector{T2})

**Sadly**, by your response, it sounds like this would *not* be an adequate solution to the "sillyness" of the @vectorize macro.

Regards,

MA

**Comment by ma-laforge** [comment]
@johnmyleswhite:

I would like to respond to your comment, but I don't think I understand.  Can you clarify?

One thing I *can* say:
There is nothing special about "Vectorizable".  The idea is that anyone can define their own function "class" (Ex: `MyFunctionGroupA<:Function`).  They could then catch calls to functions of that type by defining their own "call" signature (as demonstrated above).

Having said that: My suggestion is that functions defined within Base should use `Base.Vectorizable <: Function` (or something similar) in order to auto-generate vectorized algorithms automatically.

I would then suggest module developers implement their own functions using a pattern somewhat like:

    myfunction(x::MyType, y::MyType) <: Base.Vectorizable

Of course, they would have to provide their own version of `promote_type(::Type{MyType},::Type{MyType})` - if the default is not already to return `MyType`.

If the default vectorization algorithm is insufficient, there is nothing stopping the user from implementing their own hierachy:

    MyVectorizable{nargs} <: Function
    call(fn::MyVectorizable{2}, x, y) = ...
    
    myfunction(x::MyType, y:MyType) <: MyVectorizable{2}

MA

**Comment by johnmyleswhite** [comment]
@ma-laforge, Sorry for being unclear. My concern is that any hierarchy is always going to lack important information because Julia has single inheritance, which requires you to commit to a single parent type for each function. If you use something like `myfunction(x::MyType, y::MyType) <: Base.Vectorizable` then your function won't benefit from someone else defining a concept like `Base.NullableLiftable` that auto-generates functions of nullables.

**Comment by nalimilan** [comment]
Looks like this wouldn't be an issue with traits (cf. https://github.com/JuliaLang/julia/pull/13222). Also related is the new possibility of declaring methods as pure (https://github.com/JuliaLang/julia/pull/13555), which could automatically imply that such a method is vectorizable (at least for single-argument methods).

**Comment by ma-laforge** [comment]
@johnmyleswhite,

If I understand correctly: I don't think that is a problem for *this* case in particular.  That's because I am proposing a design pattern.  Your functions do not *have* to inherit from `Base.Vectorizable`... You can use your own.

I don't really know much about `NullableLiftables` (I don't seem to have that in my version of Julia).  However, assuming that it inherits from `Base.Function` (which is also not possible in my version of Julia):

    NullableLiftable <: Function

Your module could then implement (only once) a *new* vectorizable sub-type:

    abstract VectorizableNullableLiftable <: NullableLiftable
    
    function call{T1,T2}(fn::VectorizableNullableLiftable, v1::Vector{T1}, v2::Vector{T2})
        RT = promote_type(T1,T2) #For type stability!
        return RT[fn(v1[i],v2[i]) for i in 1:length(v1)]
    end

So, from now on, anyone that defines a function `<: VectorizableNullableLiftable` would get your vectorization code applied automatically!

    function mycooladdon(scalar1, scalar2) <: VectorizableNullableLiftable
    ...

I do understand, that having more than one Vectorizable-type is still somewhat of a pain (and a bit inelegant)...  But at least it would remove one of the annoying repetitions(1) in Julia (having to register *each* newly added function with a call to @vectorize_Xarg).

(1)**That's assuming Julia supports inheritance on functions** (ex: `myfunction(...)<: Vectorizable`) - which it doesn't appear to, on v0.4.0.  The solution I got working in Julia 0.4.0 is just a hack... You still have to register your function... not much better than calling @vectorize_Xarg

MA

**Comment by JeffBezanson** [comment]
I still think it's kind of the wrong abstraction. A function that can or should be "vectorized" is not a particular kind of function. *Every* function can be passed to `map`, giving it this behavior.

BTW, with the change I'm working on in the jb/functions branch, you will be able to do `function f(x) <: T` (though, clearly, only for the first definition of `f`).

**Comment by ma-laforge** [comment]
Ok, I think I better understand what you are looking for... and it is not what I suggested.  I think that might be part of the issues @johnmyleswhite had with my suggestions as well...

**...But if I now understand what the issue is, the solution appears even simpler to me:**

    function call{T1,T2}(fn::Function, v1::Vector{T1}, v2::Vector{T2})
        RT = promote_type(T1,T2) #For type stability!
        return RT[fn(v1[i],v2[i]) for i in 1:length(v1)]
    end
    
    myadd(x::Number, y::Number) = x+y+1

Since `myadd` is of type `Function`, it should get trapped by the `call` function... which it does:

    call(myadd,collect(1:10),collect(21:30.0)) #No problem

But `call` does not auto-dispatch on functions, for some reason (not sure why):

    myadd(collect(1:10),collect(21:30.0)) #Hmm... Julia v0.4.0 does not dispatch this to call...

**But I imagine this behaviour should not be too difficult to change**.  Personally, I don't know how I feel about making such all-encompassing catch-all functions, but it sounds like that's what you want.

**Something odd I noticed:** Julia already auto-vectorizes functions if they are not typed:

    myadd(x,y) = x+y+1 #This gets vectorized automatically, for some reason

**RE: BTW...:**
Cool!  I wonder what neat things I will be able to do by subtyping functions :).

**Comment by JeffBezanson** [comment]
> Julia already auto-vectorizes functions if they are not typed

It appears that way because the `+` operator used inside the function is vectorized.

**Comment by JeffBezanson** [comment]
> But I imagine this behaviour should not be too difficult to change. Personally, I don't know how I feel about making such all-encompassing catch-all functions, but it sounds like that's what you want.

I share your reservations. You can't sensibly have a definition that says "here's how to call any function", because each function itself says what to do when it's called --- that's what a function is!

**Comment by Ismael-VC** [comment]
> I should have said: ... user defined non-unicode infix operators, as I don't think we want to require users to type unicode characters to access such a core functionality. Although I see that $ is actually one of those added, so thank you for that! Wow, so this actually works in Julia today (even if it's not "fast"... yet):
> 
> julia> ($) = map
> julia> sin $ (0.5 * (abs2 $ (x-y)))

@binarybana how about `↦` / `\mapsto` ?

```julia
julia> x, y = rand(3), rand(3);

julia> ↦ = map    # \mapsto<TAB>
map (generic function with 39 methods)

julia> sin ↦ (0.5 * (abs2 ↦ (x-y)))
3-element Array{Float64,1}:
 0.271196
 0.0927406
 0.0632608
```

There are also:

* `⤇`	/ `\Mapsto`
* `⟼` /  `\longmapsto`
* `⟾` / `\Longmapsto`
* `⤅` / `\twoheadmapsto`
* `⊸` / `\multimap` (parallel map?)

**Comment by mlhetland** [comment]
FWIW, I'd at least initially assume that `\mapsto` was an alternate syntax for lambdas, as it is commonly used in mathematics and, essentially (in its ASCII incarnation, `->`) in Julia, too. I think this would be rather confusing.

**Comment by mlhetland** [comment]
Speaking of math … In model theory I've seen mapping used by applying a function to a tuple without parentheses. That is, if `\bar{a}=(a_1, \dots, a_n)`, then `f(\bar{a})` is `f(a_1, \dots, a_n)` (i.e., essentially, `apply`) and `f\bar{a}` is `(f(a_1), \dots, f(a_n))` (i.e., `map`). Useful syntax for defining homomorphisms, etc., but not all that easily transferable to a programming language :-}

**Comment by Ismael-VC** [comment]
How about any of the other alternatives like `\Mapsto`, would you confuse it with `=>` (Pair)? I think both symbols are distinguishable here side by side:

* `->`
* `↦`

There are lots of symbols that look alike, what would be the reason to have so many of them if we only use the ones that look **very** different or are pure ASCII?

> I think this would be rather confusing.

I think documentation and experience solves this, do you agree?

There are also lots of other arrow like symbols, I honestly don't know what are they used for in mathematics or else, I only proposed this ones because they have `map` in their names! :smile:

**Comment by mlhetland** [comment]
I guess my point is that `->` is Julia's attempt at representing `↦` in ASCII. So using `↦` to mean something other seems ill-advised. It's not that I can't tell them apart visually :-)

My gut feeling is just that if we're going to use well-established mathematical symbols, we might want to at least think through how the Julia usage differs with the established usage. Seems logical to select symbols with `map` in their names, but in this case that refers to the definition of a map (or of maps of different types, or the types of such maps). That's also true of the usage in Pair, more or less, where rather than defining a full function by defining what a parameter maps to, you actually list what a given argument (parameter value) maps to – i.e., it's an element of an explicitly enumerated function, conceptually (e.g., a dictionary).

**Comment by nalimilan** [comment]
@Ismael-VC The problem with your suggestion is that you need to call `map` twice, while the ideal solution would involve no temporary arrays and reduce to `map((a, b) -> sin(0.5 * abs2(a-b)), x, y)`. Besides, repeating `↦ ` twice isn't great both visually and for typing (for which an ASCII equivalent would be good to have).

**Comment by tkelman** [comment]
R users might dislike this idea, but if we move towards deprecating the current infix-macro special case parsing of `~` (packages like GLM and Dataframes would need to move to macro-parsing their formula DSL's, ref https://github.com/JuliaStats/GLM.jl/issues/116), that would free up the rare commodity of an infix ascii operator. `a ~ b` could be defined in base as `map(a, b)`, and maybe `a .~ b` could be defined as `broadcast(a, b)`? If it parses as a conventional infix operator then macro DSL's like an emulation of the R formula interface would be free to implement their own interpretation of the operator inside macros, as JuMP does with `<=` and `==`.

**Comment by Ismael-VC** [comment]
:+1: for less special casing and more generality and consistency.

**Comment by quinnj** [comment]
+1 Tony.

**Comment by nalimilan** [comment]
@tkelman To be clear, how would you write e.g. `sin(0.5 * abs2(a-b))` in a fully vectorized way?

**Comment by tkelman** [comment]
With a comprehension, probably. I don't think infix map would work for varargs or in-place, so the possibility of free syntax doesn't solve all issues.

**Comment by nalimilan** [comment]
So that wouldn't really fix this issue. :-/

So far the `sin(0.5 * abs2(a-b)) over (a, b)` syntax (or a variant, possibly using an infix operator) is the most appealing.

**Comment by tkelman** [comment]
The title of this issue is "Alternate syntax for `map(func, x)`". Using an infix operator does not solve map/loop fusion to eliminate temporaries but I think that may be an even broader, related but technically separate issue than the syntax.

**Comment by Ismael-VC** [comment]
Yes, I agree with @tkelman , the point is to have alternative syntax for `map`, which is why I suggested using `\mapsto`, `↦`. What @nalimilan mentions seems to be broader and better suited to another issue IMHO, `How to fully vecotrize expressions`? 

**`<rambling>`**
This issue has been going on for more than a **year** now! But we could have `Alternative syntax for map(func, x)` **now**. Out of ±450 julian contributors only 41 (that's a lot for a github issue but clearly not sufficient in this case) have been able to find this issue and/or willing to share an opinion, all in all there are not that much different suggestions (that are not just minor variations of the same concept). 

I know some of you don't like the idea or see value in making polls, but since I don't need permission from any of you, I'll just do it anyway. It's kinda sad the way we are **not** taking full leverage from our **community and social networks** and other communities also even more sad that we are not seeing the value of it, let's see if I can gather more **different** and **fresh** opinions, or at least check out what does the majority feels about the current opinions in this particular issue, as an experiment and see how it goes. Maybe it's indeed useless, maybe it isn't, there is only one way to really know.
**`</rambling>`**

**Comment by toivoh** [comment]
@Ismael-VC: If you really do want to make a poll, the first thing that you must do is to carefully consider the question that you want to ask. You can't expect everyone to read through the entire thread and summarize the options that have been up for discussion individually.

**Comment by nalimilan** [comment]
`map(func, x)` also covers things like `map(v -> sin(0.5 * abs2(v)), x)`, and this is what this thread discussed. Let's not move this to another thread, as it would make it harder to keep in mind all of the proposals discussed above.

I'm not opposed to adding syntax for the simple case of applying a generic function using `map`, but if we do it I think it would be a good idea to consider the broader picture at the same time. If it wasn't for that, the issue could have been fixed a long time ago already.

@Ismael-VC Polls are unlikely to help here IMHO. We're not trying to find out which of several solutions is the best one, but rather to find a solution which nobody has really found already. This discussion is already long and involved many people, I don't think adding more will help.

**Comment by JeffBezanson** [comment]
@Ismael-VC That's fine, feel free to make a poll. In fact I have done a couple doodle polls on issues in the past (e.g. http://doodle.com/poll/s8734pcue8yxv6t4). In my experience, the same or fewer people vote in polls than discuss in issue threads. It makes sense for very specific, often superficial/syntax issues. But how is a poll going to generate fresh ideas when all you can do is pick from existing options?

Of course the real goal of this issue is to eliminate implicitly vectorized functions. In theory, syntax for `map` is sufficient for that, because all of these functions are just doing `map` in every case.

**Comment by JeffBezanson** [comment]
I've tried to look for existing math notation for this, but you tend to see comments to the effect that the operation is too unimportant to have notation! In the case of arbitrary functions in a mathematical context I can almost believe this. However the closest thing seems to be Hadamard product notation, which has a few generalizations: https://en.wikipedia.org/wiki/Hadamard_product_(matrices)#Analogous_Operations

That leaves us with `sin∘x` as @rfourquet suggested. Doesn't seem too helpful, since it requires unicode and isn't widely known anyway.

**Comment by quinnj** [comment]
@nalimilan, I would think you would just do `sin(0.5 * abs2(a-b)) ~ (a,b)` which would translate to `map((a,b)->sin(0.5 * abs2(a-b)), (a,b))`. Not sure if that's quite right, but I think it would work.

I'm also wary of delving too much into the let-me-give-you-a-huge-complicated-expression-and-you-perfectly-auto-vectorize-it-for-me problem. I think the ultimate solution in that regard is building a full on DAG of expressions/tasks + query planning, etc. But I think that's a much bigger fish to fry than just having convenient syntax for `map`.

**Comment by nalimilan** [comment]
@quinnj Yeah, that's essentially the `over` syntax proposed above, except with an infix operator.

**Comment by johnmyleswhite** [comment]
Serious comment: I think you're likely going to re-invent SQL if you pursue this idea far enough since SQL is essentially a language for composing element-wise functions of many variables that are subsequently applied via row-wise "vectorization".

**Comment by mdcfrancis** [comment]
@johnmyleswhite agree, starts to look like a DSL aka Linq

**Comment by mdcfrancis** [comment]
On the threads posted topic, you could specialize the |> 'pipe' operator and get the map style functionality. You can read it as pipe the function to the data. As an extra bonus you can use the same to perform function composition. 

    julia> (|>)(x::Function, y...) = map(x, y... )
    |> (generic function with 8 methods)

    julia> (|>)(x::Function, y::Function) = (z...)->x(y(z...))
    |> (generic function with 8 methods)

    julia> sin |> cos |> [ 1,2,3 ]
    3-element Array{Float64,1}:
      0.514395
     -0.404239
     -0.836022

    julia> x,y = rand(3), rand(3)
    ([0.8883630054185454,0.32542923024720194,0.6022157767415313],    [0.35274912207468145,0.2331784754319688,0.9262490059844113])

    julia> sin |> ( 0.5 *( abs( x - y ) ) )
    3-element Array{Float64,1}:
     0.264617
     0.046109
     0.161309

**Comment by JeffBezanson** [comment]
@johnmyleswhite That's true, but there are worthwhile intermediate goals that are quite modest. On my branch, the `map` version of multi-operation vectorized expressions is already faster than what we have now. So figuring out how to smoothly transition to it is somewhat urgent.

**Comment by nalimilan** [comment]
@johnmyleswhite Not sure. A lot of SQL is about selecting, ordering and merging rows. Here we're only speaking about applying a function element-wise. Besides, SQL doesn't provide any syntax to distinguish reductions (e.g. `SUM`) from element-wise operations (e.g. `>`, `LN`). The latter are simply automatically vectorized just like in Julia currently.

**Comment by mlhetland** [comment]
@JeffBezanson The beauty of the use of `\circ` is that if you interpret an indexed family as a function from the index set (which is the standard mathematical "implementation"), then mapping *is* simply composition. So `(sin ∘ x)(i)=sin(x(i))`, or, rather `sin(x[i])`.

The use of the pipe, as @mdcfrancis mentions, would essentially just be "diagram order" composition, which is often done with a (possibly fat) semicolon in mathematics — but we already have the pipe operator, of course.

If neither one of these composition operators are okay, then one could use some others. For example, at least some authors use the lowly `\cdot` for abstract arrow/morphism composition, as it is essentially the "multiplication" of the groupoid (more or less) of arrows.

And if one wanted an ASCII analog: There are also authors who actually use a period to indicate multiplication. (I may have seen some use it for composition as well; can't remember.)

So one could have `sin . x` … buut I guess that would be confusing :-}

**Comment by mlhetland** [comment]
Still … that last analogy might be an argument for one of the really early proposals, i.e., `sin.(x)`. (Or maybe that's far-fetched.)

**Comment by mschauer** [comment]
Let's try it from a different angle, don't shoot me.

If we define ``..`` by ``collect(..(A,B)) == ((a[1],..., a[n]), (b[1], ...,b[n])) == zip(A,B)``, then  using ``T[x,y,z] = [T(x), T(y), T(z)]`` formally it holds that
```
map(f,A,B) = [f(a,b), ..., f(an,bn)] = f[zip(A,B)...] = f[..(A,B)]
```
That motivates at least one syntax for map which does not interfere with the syntax for array construction. With ``::`` or ``table`` the extension ``f[::(A,B)] = [f(a[i], b[j]) for i in 1:n, j in 1:n]`` this extends at least to a second interesting case.

**Comment by Ismael-VC** [comment]
> carefully consider the question that you want to ask.

@toivoh Thank you, I will. I'm currently evaluating several poll/survey software.

> find a solution which nobody has really found already

@nalimilan nobody among us, that is. :smile: 

> how is a poll going to generate fresh ideas when all you can do is pick from existing options?
> same or fewer people vote in polls than discuss in issue threads.

@JeffBezanson I'm glad to hear that you've already done polls, keep it up! 

* How have you've been promoting your polls?
* From the poll/survey software I've evaluated so far kwiksurveys.com let users add their own opinions instead of *none of this options is for me* option.

> `sin∘x` Doesn't seem too helpful, since it requires Unicode and isn't widely known anyway.

We have so many Unicode, let's use it, we even have nice way to use them with tab completion, what's wrong with using them then?, If it's not know, let's document and educate, if it doesn't exist, what's wrong with inventing it? Do we really need to wait for someone else to invent it and use it so we can take it as precedent and only after then consider it? 

`∘` has precedent, so the problem is that it's Unicode? why? when are we going to start using Unicode then? never?

By that logic, Julia isn't widely known anyway, however those who want to learn, will. It just doesn't make sense to me, in my very humble opinion.

**Comment by JeffBezanson** [comment]
Fair enough, I'm not totally against `∘`. Requiring unicode to use a pretty basic feature is just one mark against it. Not necessarily enough to sink it completely.

**Comment by mlhetland** [comment]
Would it be entirely crazy to use/overload `*` as an ASCII alternative? I'd say it could be argued that it makes some sense mathematically, but I guess it might be hard to discern its meaning at times… (Then again, if it's restricted to the `map` functionality, then `map` is already an ASCII alternative, no?)

**Comment by hayd** [comment]
> The beauty of the use of `\circ` is that if you interpret an indexed family as a function from the index set (which is the standard mathematical "implementation"), then mapping *is* simply composition.

I'm not sure I buy this.

**Comment by mlhetland** [comment]
@hayd Which part of it? That an indexed family (e.g., a sequence) can be seen as a function from the index set, or that mapping over it becomes composition? Or that this is a useful perspective in this case?

**Comment by mlhetland** [comment]
The first two (mathematical) points are pretty uncontroversial, I think. But, yeah, I'm not going to advocate strongly for using this here – it was mostly an "Ah, that kinda fits!" reaction.

**Comment by mdcfrancis** [comment]
@mlhetland |> is pretty close to -> and works today - it also has the 'advantage' of being right associative.  

    x = parse( "sin |> cos |> [1,2]" )
    :((sin |> cos) |> [1,2])

**Comment by mlhetland** [comment]
@mdcfrancis Sure. But that turns the composition interpretation I outlined on its head. That is, `sin∘x` would be equivalent to `x |> sin`, no?

**Comment by mschauer** [comment]
PS: Maybe it got lost in the "algebra", but just allowing functions in the typed array construction ``T[x,y,z]`` such that ``f[x,y,z]`` is ``[f(x),f(y),f(z)]``  gives directly
```
map(f,A) == f[A...]
```
which is quite readable and could be treated as syntax..

**Comment by JeffBezanson** [comment]
That's clever. But I suspect that if we can make it work, `sin[x...]` really loses on verbosity to `sin(x)` or `sin~x` etc.

**Comment by eschnett** [comment]
What about the syntax `[sin xs]`?

This is similar in syntax to the array comprehension `[sin(x) for x in xs]`.

**Comment by mdcfrancis** [comment]
@mlhetland sin |> x === map( sin, x )

**Comment by tkelman** [comment]
That would be the reverse order from the [current function chaining meaning](https://github.com/JuliaLang/julia/blob/763b7ebd397400e6f8224c9177d90c258efdb2f7/base/operators.jl#L204). Not that I wouldn't mind finding a better use for that operator, but would need a transition period.

**Comment by mlhetland** [comment]
@mdfrancis Yes, I get that that's what you're aiming for. Which reverses things (like @tkelman reiterates) wrt. the composition interpretation I outlined.

**Comment by bramtayl** [comment]
I think integrating vectorization and chaining would be pretty cool. I wonder if words would be the clearest operators.
Something like:

```{julia}
[1, 2] map
  +(1) map
  ^(2, _) chain
  { a = _ + 1
    b = _ - 1
    [a, b] } chain
  sum chain
  [ _, 2] chain
  reduce(+, _)
```

Several maps in a row could be combined into a single map to improve performance. Note I'm making use of R's magrittr rules for chaining (see my post in the chaining thread).

**Comment by jtravs** [comment]
I've been following this issue for a long time, and haven't commented until now, but this is beginning to get out of hand IMHO.

I strongly support the idea of a clean syntax for map. I like @tkelman 's suggestion of `~` the most as it keeps within ASCII for such basic functionality, and I quite like `sin~x`. This would allow for pretty sophisticated one-liner style mapping as discussed above. The use of `sin∘x` would also be OK. For anything more complicated I tend to think a proper loop is just much clearer (and usually the best performance). I don't really like too much 'magical' broadcasting, it makes the code much harder to follow. An explicit loop is usually clearer.

That's is not to say such functionality shouldn't be added, but let's have a nice concise `map` syntax first, especially as it's about to get super fast (from my tests of the `jb/functions` branch).

**Comment by StefanKarpinski** [comment]
Note that one of the effects of jb/functions is that `broadcast(op, x, y)` has just as good performance as the customized version `x .op y` that manually specialized the broadcasting on `op`.

**Comment by nalimilan** [comment]
> For anything more complicated I tend to think a proper loop is just much clearer (and usually the best performance). I don't really like too much 'magical' broadcasting, it makes the code much harder to follow. An explicit loop is usually clearer.

I don't agree. `exp(2 * x.^2)` is perfectly readable, and less verbose than `[exp(2 * v^2) for v in x]`. The challenge here IMHO is to avoid trapping people by letting them use the former (which allocates copies and doesn't fuse operations): for this, we need to find a syntax which would be short enough so that the slow form can be deprecated.

**Comment by bramtayl** [comment]
In retrospect, auto-broadcasting is a bit dangerous. There are several possible things you might want to do:

broadcast no arguments (chain)
broadcast just the chained argument (map)
broadcast all arguments (broadcast)

Each of the above could be modified, by:
Marking an item for broadcast (~)
Marking an item not to be broadcasted ( an extra set of [ ] )

Uniterable items should be handled automatically, disregarding syntax.

**Comment by toivoh** [comment]
Broadcasting only makes a difference when there would have been a dimension
mismatch otherwise. So when you say don't broadcast, you mean to give an
error instead if the size of the argument doesn't match up?

**Comment by mschauer** [comment]
> sin[x...] really loses on verbosity to sin(x) or sin~x etc.

Also, continuing the thought, the map is a ``sin[x...]`` less eager version on ``[f(x...)]``.
The syntax 
```
[exp(2 * (...x)^2)]
``` 
or something similar like ``[exp(2 * (x..)^2)]``  would be available and self explaining if ever real tacit function chaining is introduced.

**Comment by jtravs** [comment]
@nalimilan yes but that fits within my 'one-liner' category that I said was fine without a loop.

While we're listing all of our wishes: much more important to me would be for the results of `map` to be assignable without allocation or copying. This is another reason I will still prefer loops for performance critical code, but if this can be mitigated (#249 is not currently looking hopeful ATM) then this all becomes much more attractive.

**Comment by JeffBezanson** [comment]
>  results of map to be assignable without allocation or copying

Can you expand on this a bit? You can certainly mutate the result of `map`.

**Comment by malmaud** [comment]
I presume he means storing the output of `map` to a preallocated array.

**Comment by jtravs** [comment]
Yes, exactly. Apologies if that is already possible.

**Comment by JeffBezanson** [comment]
Ah, of course. We have `map!`, but as you observe #249 is asking for some nicer way to do that.

**Comment by nalimilan** [comment]
@jtravs I proposed a solution above with `LazyArray` (https://github.com/JuliaLang/julia/issues/8450#issuecomment-65106563), but so far the performance wasn't ideal.

**Comment by bramtayl** [comment]
@toivoh I made several edits to that post after I posted it. The question I was worried about is how to figure out which arguments to loop through and which arguments not to (so many mapall would be clearer than broadcast). I think if you are looping through more than one argument, expanding singleton dimensions to produce comparable arrays should always be done if necessary.

**Comment by jtravs** [comment]
Yes `map!` is exactly right. It would be good if any nice syntax sugar worked out here also covered that case. Could we not have that `x := ...` implicitly maps the RHS onto `x`.

**Comment by bramtayl** [comment]
I put up a package called ChainMap which integrates mapping and chaining.

Here's a short example:

```{julia]
@chain begin
  [1, 2]
  -(1)
  (_, _)
  map_all(+)
  @chain_map begin
    -(1)
    float
    ^(2, _)
  end
  begin
    a = @chain_map _ - 1
    b = @chain_map _ + 1
    [a, b]
  end
  sum
end
```

**Comment by mschauer** [comment]
I kept thinking about it and I think I finally figured out a consistent and Julian syntax for mapping arrays derived from array comprehension. The following proposal is julian because it builds on the array comprehension language which is already established. 

1. Starting from ``f[a...]`` which was actually proposed by  @Jutho, the convention
that for a vectors a, b
```
f[a...] == map(f, a[:])
f[a..., b...] == map(f, a[:], b[:])
etc
```
which does not introduce new symbols.

2.) On top of this, I would propose the introduction of one additional operator: a *shape preserving* splatting operator ``..`` (say). This is because ``...`` is a *flat* spatting operator, so ``f[a...]`` should return a vector and not an array even if ``a`` is ``n``-dimensional. If ``..`` chosen, then in this context,
```
f[a.., ] == map(f, a)
f[a.., b..] == map(f, a, b)
```
and the result inherits the shape of the arguments. Allowing broadcasting
```
f[a.., b..] == broadcast(f, a, b)
```
would allow writing thinks like
```
sum(f[v.., v'..]) == dot(v,v)
```
Heureka?

**Comment by johansigfrids** [comment]
This does not help with mapping expressions, does it? One of the advantages of the `over` syntax is how it works with expressions:
```
sin(x * (y - 2)) over x, y  == map((x, y) -> sin(x * (y - 2)), x, y) 
```

**Comment by mschauer** [comment]
Well, possibly via ``[sin(x.. * y..)]`` or ``sin[x.. * y..]`` above if you want to allow that. I like that a bit more than the over syntax because it gives a visual hint that the function operators on the elements and not on the containers.

**Comment by jtravs** [comment]
But can you not simplify that so that `x..` simply maps over `x`? So @johansigfrids example would be:
```julia
sin(x.. * (y.. - 2))  == map((x, y) -> sin(x * (y - 2)), x, y)
```

**Comment by mschauer** [comment]
@jtravs Because of scoping ( ``[println(g(x..))]`` vs ``println([g(x..)])``) and consistency ``[x..] = x``.

**Comment by mschauer** [comment]
One further possibility is to take ``x.. = x[:, 1], x[:, 2], etc.`` as partial splat of the leading subarrays (columns) and ``..y`` as partial splat of the trailing subarrays ``y.. = y[1,:], y[2,:]``. If both run over different indices this covers many interesting cases 
```
[f(v..)] == [f(v[i]) for i in 1:m ]
[v.. * v..] == [v[i] * v[i] for 1:m]
[v.. * ..v] == [v[i] * v[j] for i in 1:m, j in 1:n]
[f(..A)] == f[A[:, j], for j in 1:n]
[f(A..)] == f[A[i, :], for i in 1:m]
[dot(A.., ..A)] == [dot(A[:,i], A[j,:]) for i in 1:m, j in 1:n] == A*A
[f(..A..)] == [f(A[i,j]) for i in 1:m, j in 1:n]
[v..] == [..v] = v
[..A..] == A
```
(``v`` a Vector, ``A`` a matrix)

**Comment by JeffBezanson** [comment]
I prefer `over`, since it allows you to write an expression in normal syntax instead of introducing lots of square brackets and dots.

**Comment by mschauer** [comment]
You are right about the clutter, I think and I tried to adapt and systemize my proposal. To not overtax everybody's patience further I wrote my thoughts about maps and indices etc. in a gist https://gist.github.com/mschauer/b04e000e9d0963e40058 .

**Comment by stevengj** [comment]
After reading through this thread, my preference so far would be to have both `f.(x)` for simple things and people used to vectorized functions (the "`.` = vectorized" idiom is pretty common), and `f(x^2)-x over x` for more complicated expressions.

There are just way too many people coming from Matlab, Numpy, etcetera, to completely abandon vectorized function syntax, but telling them to add dots is not too bad.   But a good `over`-like syntax for vectorizing complex expressions into a single loop is also important.

**Comment by StefanKarpinski** [comment]
The `over` syntax really rubs me the wrong way. It just occurred to me why: it presumes that all usages of each variable in an expression are vectorized or non-vectorized, which may not be the case. For example, `log(A) .- sum(A,1)` – assume that we removed the vectorization of `log`. You also can't vectorize functions over expressions, which seems like a fairly major shortcoming what if I wanted to write `exp(log(A) .- sum(A,1))` and have the `exp` and the `log` vectorized and the `sum` not?

**Comment by stevengj** [comment]
@StefanKarpinski, then you should do either `exp.(log.(A) .- sum(A,1))` and accept the extra temporaries, or `s = sum(A, 1); exp(log(A) - s) over A`.    No matter what syntax we come up with, we're not going to cover all possible cases.

**Comment by simonster** [comment]
@stevengj I don't think the second example there works, because the `-` won't broadcast . Assuming `A` is a matrix, the output would be a matrix of single-row matrices, each of which is the log of an element of `A` minus the vector of sums along the first dimension. You'd need `broadcast((x, y)->exp(log(x)-y), A, sum(A, 1))`. But I think having a concise syntax for `map` is useful and doesn't necessarily need to be a concise syntax for `broadcast` as well.

**Comment by malmaud** [comment]
Will the functions that have historically been auto-vectorized like `sin` continue to be so with the new syntax, or would that become deprecated? I worry that even the `f.` syntax will feel like a 'gotcha' to a large swath of scientific programmers who aren't motivated by conceptual elegance arguments.

**Comment by stevengj** [comment]
My feeling is that the historically vectorized functions like `sin` should be deprecated in favor of `sin.`, but they should be deprecated quasi-permanently (as opposed to being removed entirely in the subsequent release) for the benefit of users from other scientific languages.

**Comment by stevengj** [comment]
One minor(?) problem with `f.(args...)`:  although the `object.(field)` syntax is for the most part rarely used and can probably be replaced with `getfield(object, field)` without too much pain, there are a *lot* of method definitions of the form `Base.(:+)(....) = ....`, and it would be painful to change these to `getfield`.

One workaround would be:

* define a deprecated binding `map(m::Module, s::Symbol) = getfield(m, s)` for backward compatibility
* support the syntax `Base.:+` (which currently fails) and recommend this in the deprecation warning for `Base.(:+)`

**Comment by ViralBShah** [comment]
I'd like to ask again - if this is something we can do in 0.5.0. I think it is important because of the deprecation of many vectorized constructors. I thought I would be ok with this, but I do find `map(Int32, a)`, instead of `int32(a)` a bit tedious. 

Is this basically just a matter of picking syntax at this point?

**Comment by nalimilan** [comment]
> Is this basically just a matter of picking syntax at this point?

I think @stevengj gave good arguments in favor of writing `sin.(x)` rather than `.sin(x)` in his PR https://github.com/JuliaLang/julia/pull/15032. So I'd say the path has been cleared.

I had one reservation about the fact that we don't have a solution to generalize this syntax to compound expressions efficiently yet. But I think at this stage we'd better merge this feature which covers most of the use cases rather than keep this discussion unresolved indefinitely.

**Comment by ViralBShah** [comment]
@JeffBezanson I am reverting the milestone on this to 0.5.0 in order to bring it up during a triage discussion - mainly to make sure I don't forget.

**Comment by ViralBShah** [comment]
Does #15032 also work for `call` - e.g. `Int32.(x)`?

**Comment by stevengj** [comment]
@ViralBShah, yes.   Any `f.(x...)` is transformed to `map(f, broadcast, x...)` at the syntax level, regardless of the type of `f`.

That's the main advantage of `.` over something like `f[x...]`, which is otherwise attractive (and would require no parser changes) but would only work for `f::Function`, and clashes conceptually a bit with `T[...]` array comprehensions.

**Comment by JeffBezanson** [comment]
`call` does not exist any more.

**Comment by stevengj** [comment]
(I also like @nalimilan's argument that `f.(x...)` makes `.(` the analogue of `.+` etc.)

**Comment by ViralBShah** [comment]
Yes the point wise analogy is the one I like best too. Can we go ahead and merge?

**Comment by tkelman** [comment]
Should the getfield with a module be an actual deprecation?

**Comment by stevengj** [comment]
@tkelman, as opposed to what?  However, the deprecation warning for `Base.(:+)` (i.e. literal symbol arguments) should suggest `Base.:+`, not `getfield`.

**Comment by stevengj** [comment]
@ViralBShah, was there any decision on this at Thursday's triage discussion?  #15032 is in pretty good shape for merging, I think.

**Comment by tkelman** [comment]
I think Viral missed that part of the call. My impression is that multiple people still have reservations about the aesthetics of `f.(x)` and might prefer either

1. an infix operator which would be simpler conceptually and in implementation, but we don't have any ascii ones available from what I can see. My earlier idea of deprecating the macro parsing of `~` would require work to replace in packages and it's probably too late to try to do that in this cycle.

2. or an alternative new syntax that makes it easier to fuse loops and eliminate temporaries. No other alternatives have been implemented to anywhere near the level of #15032, so it's looking like we should merge that and try it out despite remaining reservations.

**Comment by JeffBezanson** [comment]
Yes, I do have some reservations but I can't see a better option than `f.(x)` right now. It seems better than picking an arbitrary symbol like `~`, and I bet many who are used to `.*` (etc.) could even guess right away what it means.

One thing I would like to get a better sense of is whether people are ok with *replacing* existing vectorized definitions with `.(`. If people don't like it enough to do the replacement, I'd hesitate more.

**Comment by gabrielgellner** [comment]
As a user lurking on this discussion I would VERY much like to use this to replace my existing vectorized code. 

I largely use vectorization in julia for readability as loops are fast. So I like using it a lot for exp, sin, etc like has been mentioned previously. As I will already use .^, .* in such expressions adding the extra dot to sin. exp. etc feels really natural, and even more explicit, to me ... especially when I can then easily fold in my own functions with the general notation instead of mixing sin(x) and map(f, x). 

All to say, as regular user, I really, really hope this gets merged!

**Comment by diegozea** [comment]
I like more the proposed `fun[vec]` syntax than `fun.(vec)`.
What do you think about `[fun vec]`? It's like a list comprehension but with an implicit variable. It could allow to do `T[fun vec]`

**Comment by diegozea** [comment]
That syntax it's free in Julia 0.4 for vectors with length > 1:

```
julia> [sin rand(1)]
1x2 Array{Any,2}:
 sin  0.0976151

julia> [sin rand(10)]
ERROR: DimensionMismatch("mismatch in dimension 1 (expected 1 got 10)")
 in cat_t at abstractarray.jl:850
 in hcat at abstractarray.jl:875

```

**Comment by diegozea** [comment]
Something like `[fun over vec]` could be transformed at syntax level and maybe it worth to simplify  `[fun(x) for x in vec]` but isn't simpler than `map(fun,vec)`.

**Comment by diegozea** [comment]
Similar syntax to `[fun vec]`: The syntax `(fun vec)` is free and `{fun vec}` was deprecated.

```
julia> (fun vec)
ERROR: syntax: missing separator in tuple

julia> {fun vec}

WARNING: deprecated syntax "{a b ...}".
Use "Any[a b ...]" instead.
1x2 Array{Any,2}:
 fun  [0.3231600663395422,0.10208482721149204,0.7964663210635679,0.5064134055014935,0.7606900072242995,0.29583012284224064,0.5501131920491444,0.35466150455688483,0.6117729165962635,0.7138111929010424]
```

**Comment by stevengj** [comment]
@diegozea, `fun[vec]` was ruled out because it conflicts with `T[vec]`.  `(fun vec)` is basically Scheme syntax, with the multiple-arguments case presumably being `(fun vec1 vec2 ...)` ... this is pretty unlike any other Julia syntax.

Furthermore, remember that a main goal is to eventually have a syntax to replace `@vectorized` functions (so that we don't have a "blessed" subset of functions that "work on vectors"), and this means that the syntax needs to be palatable to people used to vectorized functions in Matlab, Python, etcetera.   It also needs to be easily composable for expressions like `sin(A .+ cos(B[:,1])`.   These requirements rule out a lot of the more "creative" proposals.

**Comment by diegozea** [comment]
`sin.(A .+ cos.(B[:,1]))` doesn't look so bad after all. This's going to need a good documentation.

`sin.(A .+ cos.(B[:,1]))` == `sin.(.+(A, cos.(B[:,1])))`

if `f.(x)` is documented as `.(`, similar to `.+`
 :+1:
else
deprecate `.+` (an similar functions) in favor of `+.` : `sin.(+.(A, cos.(B[:,1])))`  is more coherent
end

**Comment by stevengj** [comment]
@diegozea, #15032 already includes documentation, but any additional suggestions are welcome.

`.+` will continue to be spelled `.+`.  First, this placement of the dot is too entrenched, and there's not much to gain by changing the spelling here.  Second, as @nalimilan pointed out, you can think of `.(` as being a "vectorized function-call operator," and from this perspective the syntax is already consistent.

(Once the difficulties with type-computation in `broadcast` (#4883) are ironed out, my plan is to make another PR so that `a .⧆ b` for any operator `⧆` is just sugar for a call to `broadcast(⧆, a, b)`.  That way, we will no longer need to implement `.+` etcetera.)

**Comment by yuyichao** [comment]
> It also needs to be easily composable for expressions like `sin(A .+ cos(B[:,1]))`.

Is it possible to parse `f1.(x, f2.(y .+ z))` as `broadcast((a, b, c)->(f1(a, f2(b + c))), x, y, z)`?

**Comment by stevengj** [comment]
@yuyichao, loop fusion seems like it should be possible if the functions are marked as `@pure` and the eltypes are immutable, as I commented in #15032.   (But vectorized syntax like this is more convenience than for getting the last cycle out of critical inner loops.)

**Comment by stevengj** [comment]
Remember that the key goal here is to eliminate the need for `@vectorized` functions; this requires a syntax at least as general, nearly as convenient, and at least as fast.  It doesn't require loop fusion.

**Comment by yuyichao** [comment]
Is there any drawback if it does loop fusion too?

**Comment by stevengj** [comment]
@yuyichao, loop fusion is a much harder problem, and it's not always possible (e.g. see @StefanKarpinski's `exp(log(A) .- sum(A,1))` example above).   Holding out for that to be implemented will probably result in it *never* being implemented, in my opinion — we have to do this incrementally.  Start by exposing the user's intention.  If we can further optimize in the future, great.  If not, we still have a generalize replacement for the handful of "vectorized" functions available now.

**Comment by stevengj** [comment]
Another obstacle is that `.+` etc. is not currently exposed to the parser as a `broadcast` operation; it's just another function.   My plan is to change that (make `.+` sugar for `broadcast(+, ...)`), as noted above.   But again, it is much easier to make progress if the changes are incremental.

**Comment by yuyichao** [comment]
What I mean is that doing loop fusion by proving doing so is valid is hard, so we can let the parser do the transformation as part of the schematics. In the example above, it can be written as. `exp.(log.(A) .- sum(A,1))` and be parsed as `broadcast((x, y)->exp(log(x) - y), A, sum(A, 1))`.

It's also fine if `.+` doesn't belong to the same category yet (just like any non boardcasted function call will be put into the argument) and it's even fine if we will only do this (loop fusion) in a later version. I'm mainly asking if having such a schematics is possible (i.e. non-ambiguous) in the parser and if there's any draw back by allowing written vectorized and fuzed loop this way..

**Comment by yuyichao** [comment]
> doing loop fusion by proving doing so is valid is hard

I mean doing that in the compiler is hard (maybe not impossible), especially since the compiler needs to look into the complicated implementation of `broadcast`, unless we special case `broadcast` in the compiler, which is likely a bad idea and we should avoid it if possible...

**Comment by stevengj** [comment]
Maybe?   It doesn't seem impossible to define the `.(` syntax as "fusing" in this way, and leave it up to the caller to not use it for impure functions.   The best thing would be to try it and see if there are any hard cases, but I'm inclined to do this after the "non-fusing" PR.

**Comment by yuyichao** [comment]
> I'm inclined to do this after the "non-fusing" PR.

Totally agree, especially since `.+` isn't handled anyway.

**Comment by StefanKarpinski** [comment]
I don't want to derail this, but @yuyichao's suggestion gave me some ideas. The emphasis here is on which functions are vectorized, but that's always seems a bit misplaced to me – the real question is which variables to vectorize over, which completely determines the shape of the result. This is why I've been inclined to mark arguments for vectorization, rather than marking functions for vectorization. Marking arguments also allows for functions which vectorize over one argument but not another. That said, we can have both and this PR serves the immediate purpose of replacing the built-in vectorized functions.

**Comment by stevengj** [comment]
@StefanKarpinski, when you call `f.(args...)` or `broadcast(f, args...)`, it vectorizes over *all* the arguments.  (For this purpose, scalars are treated as 0-dimensional arrays.)   In @yuyichao's suggestion of `f.(args...)` = "fused broadcast" (which I am liking more and more), I think the fusion would "stop" at any expression that is not `g.(args...)` (to include `.+` in the future).

So, for example, `sin.(x .+ cos(x .^ sum(x.^2)))` would turn into `broadcast((x, _s_) -> sin(x + cos(x^_s_)), x, sum(broacast(x -> x^2, x)))`.   The caller would be responsible for not using `f.(args...)` in cases where fusion would screw up side effects.

Do you have an example in mind where this would not be enough?

**Comment by yuyichao** [comment]
> which I am liking more and more

I'm glad you like it. =)


Just yet another extension that probably doesn't belong to the same round, it might be possible to use `.=`, `.*=` or similar to solve the in place assignment issue (by making it distinct from the normal assignment)

**Comment by stevengj** [comment]
Yes, the lack of fusion for other operations was my main objection to `.+=` etcetera, but I think that would be resolved by having `.=` fuse with other `func.(args...)` calls.

**Comment by mschauer** [comment]
:thumbsup: There are two concept huddled together in this discussion which are in fact quite orthogonal:
the matlab'y "fused broadcasting operations" or ``x .* y .+ z`` and apl'y "maps on products and zips" like``f[product(I,J)...]``and ``f[zip(I,J)...]``. The talking past each other might have to do with that as well.

**Comment by stevengj** [comment]
@mschauer, f.(I, J) is already equivalent to `map(x -> f(x...), zip(I, J)` if `I` and `J` have the same shape. And if `I` is a row vector and `J` is a column vector, then `broadcast` indeed maps over the product set.  So I don't understand why you think they are "quite orthogonal."

**Comment by mschauer** [comment]
Orthogonal was not the right word, they are just different enough to coexist.

**Comment by stevengj** [comment]
The point is, though, that we don't need separate syntaxes for the two cases.  `func.(args...)` can support both.

**Comment by stevengj** [comment]
Once a member of the triumvirate (Stefan, Jeff, Viral) merges #15032 (which I think is merge-ready), I'll close this and file a roadmap issue to outline the remaining proposed changes (fix broadcast type-computation, deprecate `@vectorize`, turn `.op` into broadcast sugar, add syntactic "broadcast-fusion", and finally fuse with in-place assignment).

**Comment by mschauer** [comment]
Hey, I am very happy and grateful about 15032. I would not be dismissive of the discussion though. For example vectors of vectors and similar objects are still very awkward to use in julia but can sprout like weed as results of comprehensions. Good implicit notation not based on encoding iteration into singleton dimensions has the potential to ease this a lot, for example with the flexed out iterators and new generator expressions.

**Comment by stevengj** [comment]
I think this can be closed now in favor of #16285.
